Sources/Basic/SortedArray.swift

```diff
@@ -26,22 +26,43 @@ public struct SortedArray<Element>: CustomStringConvertible {
     
     /// Insert the given element, maintaining the sort order.
     public mutating func insert(_ newElement: Element) {
-        elements.append(newElement)
-        // FIXME: It is way too costly to sort again for just one element.
-        // We can use binary search to find the index for this element.
-        elements.sort(by: areInIncreasingOrder)
+        elements.insert(newElement, at: self.insertionIndex(for: newElement))
+    }
+    
+    fileprivate func insertionIndex(for element: Element) -> Index {
+        if self.isEmpty {
+            return 0
+        }
+        var (low, high) = (0, self.endIndex - 1)
+        var mid = 0
+        
+        while low < high {
+            mid = (low + high)/2
+            if areInIncreasingOrder(self[mid], element) {
+                low = mid + 1
+            } else {
+                high = mid
+            }
+        }
+        
+        if areInIncreasingOrder(element, self[low]) {
+            return low
+        }
+        
+        return self.endIndex
     }
     
+    
     /// Insert the given sequence, maintaining the sort order.
     public mutating func insert<S: Sequence>(contentsOf newElements: S) where S.Iterator.Element == Element {
         elements.append(contentsOf: newElements)
         elements.sort(by: areInIncreasingOrder)
```

Comment:
I think that might actually turn out to be slower if input sequence is large. Clients can always loop and use the single element API if they know input is going to be small. @codeOfRobin, Can you document this in these APIs?